addons/l10n_vn_edi_viettel/models/account_move.py

```diff
@@ -0,0 +1,766 @@
+            json_data=data
+        )
+        if error_message:
+            raise UserError(error_message)
+        if 'access_token' not in request_response:  # Just in case something else go wrong and it's missing the token
+            raise UserError(_('Connection to the API failed, please try again later.'))
+
+        access_token = request_response['access_token']
+        access_token_expiry = datetime.now() + relativedelta(seconds=request_response['expires_in'])
+
+        # Tokens are valid for 5 minutes. Storing it helps reduce api calls and speed up things a little bit.
+        self.company_id.write({
+            'l10n_vn_edi_token': access_token,
+            'l10n_vn_edi_token_expiry': access_token_expiry,
+        })
+
+        return request_response['access_token']
+
+    # -------------------------------------------------------------------------
+    # HELPERS
+    # -------------------------------------------------------------------------
+
+    @api.model
+    def _l10n_vn_edi_format_date(self, date):
+        """
+        All APIs for Sinvoice uses the same time format, being the current hour, minutes and seconds converted into
+        seconds since unix epoch, but formatting like milliseconds since unix epoch.
+        It means that the time will end in 000 for the milliseconds as they are not as of today used by the system.
+        """
+        return int(date.timestamp()) * 1000 if date else 0
+
+    @api.model
+    def _l10n_vn_edi_format_phone_number(self, number):
+        """
+        Simple helper that takes in a phone number and try to format it to fit sinvoice format.
+        SInvoice only allows digits, so we will remove any (, ), -, + characters.
+        """
+        # We first replace + by 00, then we remove all non digit characters.
+        number = number.replace('+', '00')
+        return re.sub(r'[^0-9]+', '', number)
+
+    def _l10n_vn_edi_get_sent_statuses(self):
+        self.ensure_one()
+        return {'sent', 'payment_state_to_update', 'canceled', 'adjusted', 'replaced'}
+
+    def _l10n_vn_edi_is_sent(self):
+        """ Small helper that returns true if self has been sent to sinvoice. """
+        self.ensure_one()
+        return self.l10n_vn_edi_invoice_state in self._l10n_vn_edi_get_sent_statuses()
+
+    def _l10n_vn_edi_is_edi_document(self):
+        """ Small helper that returns true if self is an invoice that should interact with the EDI. """
+        self.ensure_one()
+        return self.company_id.account_fiscal_country_id.code == 'VN' and self.is_sale_document()
+
+    def _l10n_vn_edi_lock_records(self):
+        """ To avoid sending multiple times the same invoice from different transactions,
+        we use this generic method to lock the records passed as parameter.
+        """
+        try:
+            with self.env.cr.savepoint(flush=False):
+                self._cr.execute(f'SELECT * FROM {self._table} WHERE id IN %s FOR UPDATE NOWAIT', [tuple(self.ids)])
```

Comment:
A generic method is coming for that soon (see https://github.com/odoo/odoo/pull/144061)

Reply:
Please ping me here or on discord when it is merged